Add a new API for setting thread names, useful for debugging/diagnostic

nashif opened the pull request:
Added k_thread_name_set() and enable thread name setting when declaring
static threads. This is enabled only when THREAD_MONITOR is used. System
threads get a name by default.

With shell sample, this will result for example with:

```

uart:~$ kernel threads
Threads:
*0x00400300 shell     :   options: 0x0 priority: -2
 0x00412040 logging   :   options: 0x0 priority: -2
 0x00401040 idle      :   options: 0x1 priority: 15
uart:~$ kernel stacks
0x00400300 shell      (real size 2044):	unused 1148	usage 896 / 2044 (43 %)
0x00412040 logging    (real size 764):	unused 532	usage 232 / 764 (30 %)
0x00401040 idle       (real size 252):	unused 64	usage 188 / 252 (74 %)

```

Fixes #10165

nashif committed:
kernel: remove unused macro parameter

Group parameter was not used anywhere.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
kernel: add a new API for setting thread names

Added k_thread_name_set() and enable thread name setting when declaring
static threads. This is enabled only when THREAD_MONITOR is used. System
threads get a name by default.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
shell: kernel: print thread name

Use new kernel API to display thread name when listing threads.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: test k_thread_name_set

Basic test for new API: k_thread_name_set.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nniranjhana commented on tests/kernel/threads/thread_apis/src/main.c:
```diff
@@ -93,6 +97,9 @@ void test_customdata_get_set_coop(void)
 
 	k_sleep(500);
 
+	int ret = strcmp(tid->name, "customdata");
```
Would we need a `k_thread_name_get()`?

AFAICS, changes LGTM, could not find anything amiss.

andrewboie replied:
thread->name will work from supervisor mode, but if we need to get these names from user mode then this API will need to be added and declared as a system call.

andrewboie commented on kernel/thread.c:
```diff
@@ -123,6 +123,26 @@ int saved_always_on = k_enable_sys_clock_always_on();
+#ifdef CONFIG_USERSPACE
+Z_SYSCALL_HANDLER(k_thread_name_set, data)
+{
+	_impl_k_thread_name_set((void *)data);
```
data is untrusted and could be something that is not a C string, or even a garbage pointer.
We also shouldn't be storing user-supplied data in the kernel object, it could be modified later unexpectedly.

You need to call z_user_string_alloc_copy() on the data parameter so that it is checked and copied into an allocation from the thread's resource pool (which must be inherited from its parent or set at some point with k_thread_resource_pool_assign() or k_thread_system_pool_assign())

Also, this API is not very flexible as it only works on the current thread, consider adding an explicit thread object argument.

nashif committed:
samples: synchronization: remove unused conf

prj_stack_guard.conf not related to this sample, we have tests other
samples for stack guard testing/demonstration.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: synchronisation: set thread names

- Use new k_thread_name_set to set thread names.
- Use board name in sample instead of architecture.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

dbkinder reviewed (approved):
+1 for doc change

pfalcon reviewed (commented):
This is very useful, I actually submitted a duplicate request for this feature: https://github.com/zephyrproject-rtos/zephyr/issues/10165 .

This PR now has conflicts, and I'm not sure if they're related, but there're some empty line missing/extra added in some places.

Beyond that, some cleanups are apparently needed, see inline comments.

Hope this can be revived now though, I'm definitely +1 for the feature.

pfalcon commented:
@andrewboie : Can you please re-review this, looks like TOCTOU'ish issues you raised were being addressed (not sure if fully though).

nashif committed:
logging: rename log_process_thread to logging

Keep the name short to make usuable when thread names are enabled.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

pfalcon commented on kernel/init.c:
```diff
@@ -334,10 +338,10 @@ static void prepare_multithreading(struct k_thread *dummy_thread)
-			  CONFIG_MAIN_THREAD_PRIORITY, K_ESSENTIAL);
 
+			  CONFIG_MAIN_THREAD_PRIORITY, K_ESSENTIAL,
+			  NULL);
```
> Why not setting main thread name here?

This was "resolved" without response.

pfalcon reviewed (changes requested):
Not all comments were responded to/addressed.

nashif committed:
shell: kernel: add more thread data to threads command

Show thread and stack data all under one command and reformat output.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

pfalcon reviewed (approved):
Thanks for fixing the main thread case!

(Now needs rebasing.)
